feat: introduce Facets- Form, Key-value, Image, and Rating Indicator

kavya-b opened the pull request:
## Related Issue
Closes SAP/fundamental-styles#2020

## Description
This PR introduces Facets. Facets are used with Object Page and Dynamic Page usually, but the facet types can also be used independently. Therefore, we are introducing Facets as a component by itself, which consists of different facet types- form facet, key-value facet, image facet, and rating indicator facet.

It also introduce a couple of new classes in Dynamic Page to work with the image facet when used in the header.

## Screenshots
![image](https://user-images.githubusercontent.com/53509521/105485640-3ca17080-5cd3-11eb-86d7-da6f65378305.png)

#### Please check whether the PR fulfills the following requirements

1. The output matches the design specs
- [x] All values are in `rem`
- [x] Text elements follow the truncation rules
- [x] hover state of the element follow design spec
- [x] focus state of the element follow design spec
- [x] active state of the element follow design spec
- [x] selected state of the element follow design spec
- [x] selected hover state of the element follow design spec
- [x] pressed state of the element follow design spec
- [x] Responsiveness rules - the component has modifier classes for all breakpoints
- [x] Includes Compact/Cosy/Tablet design
- [x] RTL support
2. The code follows fundamental-styles code standards and style
- [x] only one top level `fd-*` class is used in the file
- [x] BEM naming convention is used
- [x] Mixins are used for repeatable code (`fd-rtl`, `fd-ellipsis`, `fd-flex`, `fd-selected`, `fd-focus`, ect.)
- [x] A11y support - keyboard support, screenreader support, proper ARIA attributes, etc.
- [x] `fd-reset()` mixin is applied to all elements
- [x] Variables are used, if some value is used more than twice.
- [x] Checked if current components can be reused, instead of having new code.
3. Testing
- [x] tested Storybook examples with "CSS Resources" `normalize` option 
- [x] tested Storybook examples with "CSS Resources" `unnormalize` option 
- [x] Verified all styles in IE11
- [x] Updated tests
4. Documentation
- [x] Storybook documentation has been created/updated
- [n/a] Breaking Changes [wiki](https://github.com/SAP/fundamental-styles/wiki/Breaking-Changes) has been updated in case of breaking changes.

InnaAtanasova commented:
Hey @kavya-b I reviewed your PR but I have some points I want to discuss with my team. 
My biggest concern is that this component does not exist in Fiori 3 specs and I'm not sure we can introduce it in Fundamentals. It should prob go to thee `Layouts` section as it's not really a Pattern. 
The other thing I would improve is the structure of the classes. 
Your block element is `.fd-facet`. Then you can have a `.fd-facet__container` class where you bring the reset and the margins bottom and right. You can have a modifier class to overwrite the defaults depending on the element. For example, for rating indicator you will have:
```
.fd-facet
    .fd-facet__container.fd-facet__container--rating-indicator
```

For form it will be:
```
.fd-facet
    .fd-facet__container.fd-facet__container--form
```

InnaAtanasova commented on src/facet.scss:
```diff
@@ -0,0 +1,171 @@
+  }
+
+  &__rating-indicator {
+    @include fd-reset();
```
This is used like this `fd-rating-indicator fd-facet__rating-indicator` and if you reset it here, you will reset the styling coming from `fd-rating-indicator`

InnaAtanasova commented on stories/facets/facets.stories.js:
```diff
@@ -0,0 +1,819 @@
+        <div class="fd-facet__rating-indicator-container" role="group" aria-labelledby="ratingFacetTitle1">
+            <div class="fd-facet__title-container">
+                <h1 class="fd-facet__title" id="ratingFacetTitle1">Average User Rating</h1>
+                <h2 class="fd-facet__subtitle fd-form-label">6 reviews</h2>
```
+1 to this comment. There's no `fd-facet__subtitle` in the facets scss file.

kavya-b replied:
thanks for noticing, this was not needed. 😅

kavya-b commented:
> Hey @kavya-b I wrote to the designer about this "component". Our next sync up meeting with them is on Feb.2nd. We discussed it today during our team meeting and we think it can remain a stand-alone component but it should be hidden as Numeric Content and Object Attribute and you can provide examples of it in Object Page and/or Dynamic Page. More information I will provide after the meeting with the designers.

I see. From what I understand now, we will continue having Facets as a component(incorporating above comments on the structure as well), but will not show it in the documentation pages. Instead, we would include it directly in the Dynamic Page example similar to [this](https://deploy-preview-2077--fundamental-styles.netlify.app/?path=/docs/components-facets--combined#facets-in-object-page). Is my understanding correct? Can I start working on making these changes?

cc: @manjunathanagaraj

kavya-b commented:
> The other thing I would improve is the structure of the classes.
> Your block element is `.fd-facet`. Then you can have a `.fd-facet__container` class where you bring the reset and the margins bottom and right. You can have a modifier class to overwrite the defaults depending on the element. For example, for rating indicator you will have:
> 
> ```
> .fd-facet
>     .fd-facet__container.fd-facet__container--rating-indicator
> ```
> 
> For form it will be:
> 
> ```
> .fd-facet
>     .fd-facet__container.fd-facet__container--form
> ```
> 
> This will make the angular implementation easier as the type of the component can be specified with an input parameter.

I was looking into this and I'm not sure I understand how much easier implementation would get making this change. Sure we can take `type` from user, say `'rating-indicator'`, but within that, we have the title, the subtitle/label, and then the actual rating indicator. And this syntax of title->label->content format is not the same across all facets. For example, for key-value facet, we only have a title and a value; and a form facet has title, label and corresponding value. So even after we take `type` from user, we would still need more specific information to map to the correct facet, right?

InnaAtanasova commented:
> > Hey @kavya-b I wrote to the designer about this "component". Our next sync up meeting with them is on Feb.2nd. We discussed it today during our team meeting and we think it can remain a stand-alone component but it should be hidden as Numeric Content and Object Attribute and you can provide examples of it in Object Page and/or Dynamic Page. More information I will provide after the meeting with the designers.
> 
> I see. From what I understand now, we will continue having Facets as a component(incorporating above comments on the structure as well), but will not show it in the documentation pages. Instead, we would include it directly in the Dynamic Page example similar to [this](https://deploy-preview-2077--fundamental-styles.netlify.app/?path=/docs/components-facets--combined#facets-in-object-page). Is my understanding correct? Can I start working on making these changes?
> 
> cc: @manjunathanagaraj

@kavya-b We have exactly the same case with Card component and Numeric content component. You can check [here](https://deploy-preview-2077--fundamental-styles.netlify.app/?path=/docs/components-card--card-anatomy#analytical-card). Numeric content is used in multiple other components but is not a stand-alone component and that's why is hidden from the documentation and examples of its usage are shown in Cards, Tiles, etc.

InnaAtanasova commented:
> > The other thing I would improve is the structure of the classes.
> > Your block element is `.fd-facet`. Then you can have a `.fd-facet__container` class where you bring the reset and the margins bottom and right. You can have a modifier class to overwrite the defaults depending on the element. For example, for rating indicator you will have:
> > ```
> > .fd-facet
> >     .fd-facet__container.fd-facet__container--rating-indicator
> > ```
> > 
> > 
> > For form it will be:
> > ```
> > .fd-facet
> >     .fd-facet__container.fd-facet__container--form
> > ```
> > 
> > 
> > This will make the angular implementation easier as the type of the component can be specified with an input parameter.
> 
> I was looking into this and I'm not sure I understand how much easier implementation would get making this change. Sure we can take `type` from user, say `'rating-indicator'`, but within that, we have the title, the subtitle/label, and then the actual rating indicator. And this syntax of title->label->content format is not the same across all facets. For example, for key-value facet, we only have a title and a value; and a form facet has title, label and corresponding value. So even after we take `type` from user, we would still need more specific information to map to the correct facet, right?

I don't know the details of the component but here's how I would approach it (this is just my point of view, don't insist on you agreeing with me :) )
I would make "universal classes" for the facet component:

```
.fd-facet
    .fd-facet__container
        .fd-facet__title
        .fd-facet__subtitle
        .fd-facet__header
        .fd-facet__label
        .fd-facet__content
        .fd-facet__something-else ......
```

There are would put the repeating (shared) css rules that are valid in all cases. Then I would add the modifier classes:
```
.fd-facet.fd-facet--rating-indicator
.fd-facet.fd-facet--form
.fd-facet.fd-facet--key-value
.fd-facet.fd-facet--something-else ......
```

and from there, based on the modifier class I would change the "children" of the facet:
```
.fd-facet.fd-facet--rating-indicator {
     .fd-facet__container {
         // code specific for the container of the rating indicator
     }

     .fd-facet__subtitle {
         // code specific for the subtitle of the rating indicator
     }
}
```
and
```
.fd-facet.fd-facet--form {
     .fd-facet__container {
         // code specific for the container of the form
     }

     .fd-facet__subtitle {
         // code specific for the subtitle of the form
     }
}
```

And it's ok if you have for example `.fd-facet__header` that is only part of rating indicator, the child doesn't have to be present in all components, but at least you can have a component structure that can live independently of other components and use the modifier classes for the specific needs.

That's how I would approach this component. Again, just a suggestion.

kavya-b commented:
> > > The other thing I would improve is the structure of the classes.
> > > Your block element is `.fd-facet`. Then you can have a `.fd-facet__container` class where you bring the reset and the margins bottom and right. You can have a modifier class to overwrite the defaults depending on the element. For example, for rating indicator you will have:
> > > ```
> > > .fd-facet
> > >     .fd-facet__container.fd-facet__container--rating-indicator
> > > ```
> > > 
> > > 
> > > For form it will be:
> > > ```
> > > .fd-facet
> > >     .fd-facet__container.fd-facet__container--form
> > > ```
> > > 
> > > 
> > > This will make the angular implementation easier as the type of the component can be specified with an input parameter.
> > 
> > 
> > I was looking into this and I'm not sure I understand how much easier implementation would get making this change. Sure we can take `type` from user, say `'rating-indicator'`, but within that, we have the title, the subtitle/label, and then the actual rating indicator. And this syntax of title->label->content format is not the same across all facets. For example, for key-value facet, we only have a title and a value; and a form facet has title, label and corresponding value. So even after we take `type` from user, we would still need more specific information to map to the correct facet, right?
> 
> I don't know the details of the component but here's how I would approach it (this is just my point of view, don't insist on you agreeing with me :) )
> I would make "universal classes" for the facet component:
> 
> ```
> .fd-facet
>     .fd-facet__container
>         .fd-facet__title
>         .fd-facet__subtitle
>         .fd-facet__header
>         .fd-facet__label
>         .fd-facet__content
>         .fd-facet__something-else ......
> ```
> 
> There are would put the repeating (shared) css rules that are valid in all cases. Then I would add the modifier classes:
> 
> ```
> .fd-facet.fd-facet--rating-indicator
> .fd-facet.fd-facet--form
> .fd-facet.fd-facet--key-value
> .fd-facet.fd-facet--something-else ......
> ```
> 
> and from there, based on the modifier class I would change the "children" of the facet:
> 
> ```
> .fd-facet.fd-facet--rating-indicator {
>      .fd-facet__container {
>          // code specific for the container of the rating indicator
>      }
> 
>      .fd-facet__subtitle {
>          // code specific for the subtitle of the rating indicator
>      }
> }
> ```
> 
> and
> 
> ```
> .fd-facet.fd-facet--form {
>      .fd-facet__container {
>          // code specific for the container of the form
>      }
> 
>      .fd-facet__subtitle {
>          // code specific for the subtitle of the form
>      }
> }
> ```
> 
> And it's ok if you have for example `.fd-facet__header` that is only part of rating indicator, the child doesn't have to be present in all components, but at least you can have a component structure that can live independently of other components and use the modifier classes for the specific needs.
> 
> That's how I would approach this component. Again, just a suggestion.

Thanks for your detailed feedback, @InnaAtanasova ! I came up with the following modified structure after taking some of your suggestions: 
![image](https://user-images.githubusercontent.com/53509521/106167066-d6c15700-61b2-11eb-8006-cd602fcc96c9.png)
so, we have 
```
.fd-facet //styles for when many facets are used together
     .fd-facet-container //common margins and fd-reset() for each facet container
        modifiers for .fd-facet-container:
        .fd-facet-container--form  //form-facet-specific container level styles
        .fd-facet-container--image //image-facet-specific container level styles
        .fd-facet-container--key-value  //key-value-facet-specific container level styles
        .fd-facet-container--rating-indicator  //rating-indicator-facet-specific container level styles

          .fd-facet__title-container //title and subtitle common classes
        .fd-facet__content-container with modifiers for fd-facet__content-container--form and fd-facet__content-container--rating-indicator since their contents also have specific styles.
       
```
We also have other classes like `.fd-facet__form-value, .fd-facet__object-status, .fd-facet__object-number, fd-facet__rating-container, fd-facet__rating-dynamic-text, .fd-facet-container--image-header ` that deal with very specific style changes(like style the form facet's value, or modify object status/number/rating indicator/image in object page header) and only take affect when used in the respective facet's context.
Let me know what you think; will push in the changes tomorrow.

kavya-b commented:
@InnaAtanasova @JKMarkowski 
- have addressed your comments in addition to changing the structure of the classes. 
- have also made Facets 'Dev' only so the stories won't be visible anymore on the documentation page. 
- have added example usage of facets in Dynamic Page documentation page.
Please check.

InnaAtanasova reviewed (commented):
1. The facet containers are all the same height. 
Fundamentals:
<img width="819" alt="Screen Shot 2021-02-03 at 9 12 40 AM" src="https://user-images.githubusercontent.com/39598672/106758955-257e4f00-6600-11eb-9215-4dfd06068a63.png">
<img width="1250" alt="Screen Shot 2021-02-03 at 9 12 29 AM" src="https://user-images.githubusercontent.com/39598672/106758956-2616e580-6600-11eb-8ba1-dc0da4fe9a8b.png">

Fiori:
<img width="709" alt="Screen Shot 2021-02-03 at 9 13 24 AM" src="https://user-images.githubusercontent.com/39598672/106758971-2b743000-6600-11eb-8094-6646505b93e1.png">

2. Image/Avatar container has margin bottom, in Fiori only margin right is specified.
3. `fd-facet__form-value` is not specific to facets, it says in the specs it should follow Medium Text
4. I don't understand why there's a `fd-facet__title-container` and `fd-facet__title`. This can be a fundamentals title component with font-size of 1rem: `h5 class="fd-title fd-title--h5` and then add a modifier class to set the spacing

I see facets as containers that hold existing components (none of the components are specific to Facets) and just add spacings to these components. 
I will ask Manju to discuss this component during one of the components review meetings this week or next one.

kavya-b commented on src/facet.scss:
```diff
@@ -0,0 +1,167 @@
+    &--key-value {
+      .#{$block}__object-status {
+        font-family: var(--sapFontFamily);
+        font-size: $fd-object-status-facet-font-size;
```
Object Status Large design does have 1.5rem font-size but the font-family and how the status is displayed with the label(horizontal or vertical) is also different:
![image](https://user-images.githubusercontent.com/53509521/106841892-ff70b180-66c8-11eb-9b4a-f9971a88578d.png)

![image](https://user-images.githubusercontent.com/53509521/106841810-cd5f4f80-66c8-11eb-8a80-3df73b6cd0a8.png)

the reason why extra styles were added to object number is because when used as is, the alignment was looking off:
![image](https://user-images.githubusercontent.com/53509521/106841466-2ed2ee80-66c8-11eb-8d6d-50298e965e43.png)

kavya-b commented:
> 1. The facet containers are all the same height.
>    Fundamentals:
> 
> <img alt="Screen Shot 2021-02-03 at 9 12 40 AM" width="819" src="https://user-images.githubusercontent.com/39598672/106758955-257e4f00-6600-11eb-9215-4dfd06068a63.png">
> 
> <img alt="Screen Shot 2021-02-03 at 9 12 29 AM" width="1250" src="https://user-images.githubusercontent.com/39598672/106758956-2616e580-6600-11eb-8ba1-dc0da4fe9a8b.png">
> 
> Fiori:
> <img alt="Screen Shot 2021-02-03 at 9 13 24 AM" width="709" src="https://user-images.githubusercontent.com/39598672/106758971-2b743000-6600-11eb-8094-6646505b93e1.png">
> 
Will check this.

> 1. Image/Avatar container has margin bottom, in Fiori only margin right is specified.

By itself, it should not have a margin bottom, but when used in Object/Dynamic page, there is a margin bottom specified:
![image](https://user-images.githubusercontent.com/53509521/106775162-de787400-6668-11eb-8d66-b8c34105793d.png)

> 2. `fd-facet__form-value` is not specific to facets, it says in the specs it should follow Medium Text

It does inherit from Medium text, but has overriding styles for the color:
![image](https://user-images.githubusercontent.com/53509521/106775403-17b0e400-6669-11eb-991a-8b49de02929d.png)
spec:
![image](https://user-images.githubusercontent.com/53509521/106776002-ac1b4680-6669-11eb-86d5-5a4bfd322f63.png)

> 3. I don't understand why there's a `fd-facet__title-container` and `fd-facet__title`. This can be a fundamentals title component with font-size of 1rem: `h5 class="fd-title fd-title--h5` and then add a modifier class to set the spacing
> 
 `fd-facet__title-container` was added mainly to support the subtitle for rating indicator facet. Sure I can remove the `__title` element and use the title component with a modifier(modifier to title component like `fd-title--facet-spacing`?) but I think I would still need the `__title-container` element for spacing the rating indicator facet and setting the flex-direction on it:
![image](https://user-images.githubusercontent.com/53509521/106777403-f94be800-666a-11eb-8f9d-310fdd784ec8.png)

> I see facets as containers that hold existing components (none of the components are specific to Facets) and just add spacings to these components.
> I will ask Manju to discuss this component during one of the components review meetings this week or next one.

kavya-b commented:
1.  After incorporating margins/paddings and aligning to the new dynamic page structure, the facet structure has now changed/minimized to this:

- `fd-facet-group`
  -   `fd-facet fd-facet(--image or --form or --key-value or --rating-indicator)`
      -   `fd-facet__container` to give some styles to the content of a facet, like adding the correct `flex-direction`
          -    `fd-facet__rating-container` and `fd-facet__rating-dynamic-text` - related to alignment in rating indicator facet


2. Removed `fd-facet__title` and replaced with `fd-title` instead since the font-size values `--sapFontLargeSize` and `--sapFontHeader5Size` evaluate to the same value.
3. Removed `fd-facet__form-value` and used `fd-text` instead since the colors `--sapTextColor` and `--sapGroup_TitleTextColor` resolve to the same value. Please let me know if I should match the variable name too; in that case I would need a separate class for this.
4. Image facet by itself should not have a margin bottom, but when used in Object/Dynamic page, there is a margin bottom specified:
![image](https://user-images.githubusercontent.com/53509521/106775162-de787400-6668-11eb-8d66-b8c34105793d.png) 
That's why you will see a bottom margin in the documentation.

@InnaAtanasova @JKMarkowski please take a look again.

JKMarkowski commented on src/facet.scss:
```diff
@@ -0,0 +1,71 @@
+    @include fd-flex(column);
+  }
+
+  &--key-value {
+    .#{$block}__object-status {
+      .#{$block}__object-status-text {
+        font-family: var(--sapFontFamily);
+      }
+    }
```
Can this be simplified to
```suggestion
  &--key-value {
      .#{$block}__object-status-text {
        font-family: var(--sapFontFamily);
      }
```
?

kavya-b replied:
Unfortunately, no. I had tried this, but object status component takes higher specificity and its `font-family` gets applied instead.

JKMarkowski commented on src/facet.scss:
```diff
@@ -0,0 +1,71 @@
+    align-items: flex-start;
+    flex-direction: column;
+
+    .#{$block}__container {
+      flex-direction: column;
+
+      .#{$block}__rating-container {
+        align-self: flex-start;
+      }
+
+      .#{$block}__rating-dynamic-text {
+        align-self: flex-end;
+      }
+    }
+  }
```
Can this structure be flatted? 
```suggestion
    .#{$block}__container {
      flex-direction: column;
      }

      .#{$block}__rating-container {
        align-self: flex-start;
      }

      .#{$block}__rating-dynamic-text {
        align-self: flex-end;
      }
  }
```

kavya-b replied:
fixed as suggested, please take a look @JKMarkowski , thanks!

JKMarkowski reviewed (commented):
Hi @kavya-b I added 2 comments, rest LGTM, let me know what you think about it, or whether it's addressed, so I will approve PR

JKMarkowski reviewed (approved):
LGTM
